engines/hadesch/hotzone.cpp

```diff
@@ -0,0 +1,205 @@
+
+namespace Hadesch {
+bool HotZone::isEnabled() const {
+	return _enabled;
+}
+
+void HotZone::setEnabled(bool val) {
+	_enabled = val;
+}
+
+
+void HotZone::setOffset(Common::Point offset)
+{
+	_offset = offset;
+}
+
+// Following function copied from sword 2.5 and simplified
+bool HotZone::isInside(const Common::Point &point_in) const {
+	Common::Point point = point_in - _offset;
+	int rcross = 0; // Number of right-side overlaps
+
+	// Each edge is checked whether it cuts the outgoing stream from the point
+	for (unsigned i = 0; i < _polygon.size(); i++) {
+		const Common::Point &edgeStart = _polygon[i];
+		const Common::Point &edgeEnd = _polygon[(i + 1) % _polygon.size()];
+
+		// A vertex is a point? Then it lies on one edge of the polygon
+		if (point == edgeStart)
+			return true;
+
+		if ((edgeStart.y > point.y) != (edgeEnd.y > point.y)) {
+			int term1 = (edgeStart.x - point.x) * (edgeEnd.y - point.y) - (edgeEnd.x - point.x) * (edgeStart.y - point.y);
+			int term2 = (edgeEnd.y - point.y) - (edgeStart.y - edgeEnd.y);
+			if ((term1 > 0) == (term2 >= 0))
+				rcross++;
+		}
+	}
+
+	// The point is strictly inside the polygon if and only if the number of overlaps is odd
+	return ((rcross % 2) == 1);
+}
+
+const Common::String &HotZone::getID() const {
+	return _hotid;
+}
+
+int HotZone::getICSH() const {
+	return _icsh;
+}
+
+HotZone::HotZone(Common::Array<Common::Point> polygon,
+		 Common::String hotid, bool enabled,
+		 int icsh) {
+	_polygon = polygon;
+	_hotid = hotid;
+	_enabled = enabled;
+	_icsh = icsh;
+}
+
+void HotZoneArray::readHotzones(Common::SharedPtr<Common::SeekableReadStream> hzFile, bool enable, Common::Point offset) {
+	if (!hzFile) {
+		debug("Invalid hzFile");
+		return;
+	}
+	TagFile tf;
+	tf.openStoreHot(hzFile);
+	Common::ScopedPtr<Common::SeekableReadStream> tcshStream(tf.getFileStream(MKTAG ('T', 'C', 'S', 'H')));
+	int hzCnt = tcshStream->readUint32LE();
+	
+	for (int idx = 0; idx < hzCnt; idx++) {
+		Common::SharedPtr<Common::SeekableReadStream> tdshFile(tf.getFileStream(MKTAG ('T', 'D', 'S', 'H'), idx));
+		TagFile tdsh;
+		tdsh.openStoreHotSub(tdshFile);
+
+		Common::SharedPtr<Common::SeekableReadStream> tvshFile(tdsh.getFileStream(MKTAG ('T', 'V', 'S', 'H')));
+
+		Common::Array<Common::Point> polygon;
+
+		for (int j = 0; j < tvshFile->size() / 8; j++) {
+			uint32 x = tvshFile->readUint32LE();
+			uint32 y = tvshFile->readUint32LE();
+			polygon.push_back(Common::Point(x, y) + offset);
+		}
+
+		Common::SharedPtr<Common::SeekableReadStream> mnshFile(tdsh.getFileStream(MKTAG ('M', 'N', 'S', 'H')));
+
+		int mnshSize = mnshFile->size();
+		char *name = new (std::nothrow) char[mnshSize + 1];
+		mnshFile->read(name, mnshSize);
+		name[mnshSize] = 0;
+
+		Common::SharedPtr<Common::SeekableReadStream> icshFile(tdsh.getFileStream(MKTAG ('I', 'C', 'S', 'H')));
+
+		int icsh = icshFile->readUint32LE();
+		
+		_hotZones.push_back(HotZone(polygon, name, enable, icsh));
+
+		delete [] name;		
```

Comment:
delete[] without a space before the square brackets

Reply:
Done